Shopify/cli
# Fix hover background for Safari in the dev console

### WHY are these changes introduced?

[Fixes #1630](https://github.com/Shopify/ui-extensions-private/issues/1630)

The hover effect for an extension row doesn't span across the whole row on Safari
<img width="1192" alt="Screen Shot 2022-08-26 at 13 41 41" src="https://user-images.githubusercontent.com/3555826/186979678-696fa301-314c-496d-8f9e-582b7a5b54a1.png">

### WHAT is this pull request doing?

It's impossible to apply a background style to a table row in Safari and the styles are instead applied to the cells. This caused the issue above since it applied a background image which had rounded corners and started/ended before the end of the full width to each cells individually.

Steps to solve the problem:
- Apply the background image to each cell so every browsers would behave the same
- Remove the rounded corners and spacing from the image of the cells
- Apply specific rounded corners to the first and last cells to achieve the expected visual

Explanation on how we got to this solution and why it's so complex at the bottom.

### Measuring impact

How do we know this change was effective? Please choose one:


### How to test your changes?

- run `yarn shopify app scaffold extension --path app-name`
- Choose `subscription-ui`
- then run `yarn shopify app scaffold extension --path integrated-conversion-app`
- visit the preview ngrok link generated for your subscription UI extension
- Hovering the dev console extension should show you the proper hover state in any browser

**Expected hover state**
<img width="1512" alt="Screen Shot 2022-08-26 at 09 45 50" src="https://user-images.githubusercontent.com/3555826/186981496-fe042123-f7e6-486c-8d8e-4a99d326b931.png">

### Why is this so complex and what was tried to get to this solution?

#### Why do we need a background-image svg?
- It's extremely difficult to style html tables and a lot of "normal" CSS properties do not apply the same or at all to table elements.
- The design requirements are to have an hover effect that has rounded corners and space between the row's borders

#### What did we try before getting to this?
##### Using a background color and creating a spacing using margin, padding at various levels and a border with the same color as the background of the console on the row
- Padding wouldn't apply to the row and making it useless for this purpose as giving padding to the cell still applies the background to the padded area
- Margin doesn't work on cell or row elements
- Adding a large border the same color as the background of the console removes the border radius on the first and last cells
##### Using pseudo elements
- Row elements do not handle `position: relative` consistently across browsers
- Managed to get it to work on Chrome, but not on Safari
##### Adding a ref to the rows in React and getting their position to then add a new element outside of the table to give it the right position using javascript
- Incredibly overkill
- Poor performance
- Did not manage to show the content over the element in Safari resulting in the background color being displayed over the text.
